### PR TITLE
Task 0: Added node taints labels and feature flags

### LINKED ISSUE #42001: NodeController should add NoSchedule taints and we should get rid of getNodeConditionPredicate()
getNodeConditionPredicate() in plugin/pkg/scheduler/factory/factory.go makes our code hard to understand because it hides the node-condition-based filtering in the node lister, which is totally non-obvious.

We should get rid of this function and have NodeController add NoSchedule taints for these situations instead. (Alas, we might not actually be able to get rid of getNodeConditionPredicate() completely until we get rid of the Unschedulable field of PodSpec, but at least we can get rid of all the other code in this function.) If there are pods that should still be able to schedule in any of these situations (e.g. DaemonSet pods?) we should add tolerations in admission control for them (e.g. see pkg/controller/daemon/daemoncontroller.go).

cc/ @kubernetes/sig-scheduling-misc  @kubernetes/sig-cluster-lifecycle-misc 
cc/ @gmarek @kevin-wangzefeng 

**Sub-tasks according to the design doc:**

- [x] Add node taints label and feature flag (#49547)
- [x] In the node controller, taint Nodes according to the Node Condition (#49257)
- [x] In DaemonSet, update DaemonSetController to Tolerant new Taints (#50186)
- [x] In admissionController, add MemoryPressure/DiskPressure toleration for no BestEffort pod (#50180)
- [x] In scheduler, disable the current behaviour of filtering out Nodes. Instead, pods will not be scheduled to tainted nodes if a toleration does not exist in its PodSpec (#50185)
- [x] Add e2e test for this feature (https://github.com/kubernetes/kubernetes/pull/52723)
- [x] Update doc for this feature (https://github.com/kubernetes/kubernetes.github.io/pull/5352)

### PR DESCRIPTION
**Pull request by k82cn**
**What this PR does / why we need it**:
Added node taint const for node condition.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: part of #42001

**Release note**:
```release-note
None
```

### CODE REVIEW

File: plugin/pkg/scheduler/algorithm/well_known_labels.go [hunk 1]
```diff
@@ -28,6 +28,26 @@ const (
 	// and removed when node becomes reachable (NodeReady status ConditionTrue).
 	TaintNodeUnreachable = "node.alpha.kubernetes.io/unreachable"
 
+	// When feature-gate for TaintBasedEvictions=true flag is enabled,
+	// TaintNodeOutOfDisk would be automatically added by node controller
+	// when node becomes out of disk, and removed when node has enough disk.
+	TaintNodeOutOfDisk = "node.alpha.kubernetes.io/outOfDisk"
```
Comment (liggitt) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: What is the rationale behind including alpha in the taint name? Is there concern a taint named `node.kubernetes.io/outOfDisk` would want to be used in some other way in the future? If not, I'd prefer taking the approach recommended for API fields:

1. Carefully choose names
2. Feature gate the setting of the data while in alpha
3. On graduation to beta, the feature gate can be enabled by default

That means that if the feature survives, clusters that enabled and used the taints in alpha state can simply continue to do so.

cc @kubernetes/sig-api-machinery-pr-reviews

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: Seems `node.kubernetes.io/outOfDisk` is better to keep the backward compatibility. @gmarek , WDYT?

Comment (wanghaoran1988) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: +1 for `node.kubernetes.io/outOfDisk`

Comment (gmarek) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: +1

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: Great! Updated to `node.kubernetes.io/xxxx`. Will handle `notReady` and `unreachable` in beta graduation PR.

File: plugin/pkg/scheduler/algorithm/well_known_labels.go [hunk 2]
```diff
@@ -28,6 +28,26 @@ const (
 	// and removed when node becomes reachable (NodeReady status ConditionTrue).
 	TaintNodeUnreachable = "node.alpha.kubernetes.io/unreachable"
 
+	// When feature-gate for TaintBasedEvictions=true flag is enabled,
+	// TaintNodeOutOfDisk would be automatically added by node controller
+	// when node becomes out of disk, and removed when node has enough disk.
+	TaintNodeOutOfDisk = "node.kubernetes.io/outOfDisk"
```
Comment (timothysc) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: these should all have .alpha.

Comment (liggitt) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: see the discussion on including alpha in the name in https://github.com/kubernetes/community/pull/819#discussion_r128887276

tl;dr: if you include alpha in the name, there is no reasonable way to migrate people to the beta and ga versions. feature-gating the alpha function is sufficient to keep it from affecting anything until it graduates to beta or an admin opts in to using it in alpha mode.

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: Update the link to open comments directly (https://github.com/kubernetes/community/pull/819/files/1bb113172319e32a8bde63f996c0b382e2fa511a#r128887276)

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: @timothysc , is that OK for your?

**Comment by k82cn** [comment]
if not comments on the name of label, can we get this PR merged firstly?

File: pkg/features/kube_features.go [hunk 3]
```diff
@@ -126,6 +126,13 @@ const (
 	//
 	// Enable equivalence class cache for scheduler.
 	EnableEquivalenceClassCache utilfeature.Feature = "EnableEquivalenceClassCache"
+
+	// owner: @k82cn
+	// alpha: v1.8
+	//
+	// Taint nodes based on their condition status for 'NetworkUnavailable',
+	// 'MemoryPressure', 'OutOfDisk' and 'DiskPressure'.
+	TaintNodesByCondition utilfeature.Feature = "TaintNodesByCondition"
```
Comment (timothysc) [inline pkg/features/kube_features.go]: Don't feature gates need to be plumbed through api-machinery to hit all component configs?

File: plugin/pkg/scheduler/algorithm/well_known_labels.go (hunk 2 above)
Comment (timothysc) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: We should update other taints then, and have a pattern for upgrade.  
/cc @davidopp

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: I'd like to update other taint when graduating from alpha to beta in separate PR.

File: pkg/features/kube_features.go (hunk 3 above)
Comment (k82cn) [inline pkg/features/kube_features.go]: Nop; AFAIK, feature gates are local map in each component for now :).

**Comment by k82cn** [comment]
@timothysc @liggitt @gmarek , can we get this merged?

**Comment by liggitt** [comment]
> @timothysc @liggitt @gmarek , can we get this merged?

My main interest was the taint name and plans for progressing to beta and stable, so no objections from me

**Comment by k82cn** [comment]
@davidopp , do you have any comments for the name of taint labels?

File: plugin/pkg/scheduler/algorithm/well_known_labels.go (hunk 2 above)
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: these look OK to me, per @liggit's comment
